arrows/vxl/high_pass_filter.cxx

```diff
@@ -0,0 +1,478 @@
+// This file is part of KWIVER, and is distributed under the
+// OSI-approved BSD 3-Clause License. See top-level LICENSE file or
+// https://github.com/Kitware/kwiver/blob/master/LICENSE for details.
+
+#include "high_pass_filter.h"
+
+#include <arrows/vxl/image_container.h>
+
+#include <vil/vil_image_view.h>
+#include <vil/vil_math.h>
+#include <vil/vil_plane.h>
+#include <vil/vil_transpose.h>
+
+namespace kwiver {
+
+namespace arrows {
+
+namespace vxl {
+
+// --------------------------------------------------------------------------------------
+/// Private implementation class
+class high_pass_filter::priv
+{
+public:
+
+  priv( high_pass_filter* parent )
+    : p( parent )
+  {
+  }
+
+  ~priv()
+  {
+  }
+
+  // Internal parameters/settings
+  std::string mode = "box";
+  size_t kernel_width = 7;
+  size_t kernel_height = 7;
+  bool treat_as_interlaced = false;
+  bool output_net_only = false;
+  high_pass_filter* p;
+
+  // Perform box filtering
+  template < typename PixType >
+  vil_image_view< PixType >
+  box_high_pass_filter( const vil_image_view< PixType >& grey_img )
+  {
+    vil_image_view< PixType > output;
+    output.set_size( grey_img.ni(), grey_img.nj(), 3 );
+
+    vil_image_view< PixType > filter_x = vil_plane( output, 0 );
+    vil_image_view< PixType > filter_y = vil_plane( output, 1 );
+    vil_image_view< PixType > filter_xy = vil_plane( output, 2 );
+
+    box_average_vertical( grey_img, filter_x, kernel_height );
+    box_average_horizontal( grey_img, filter_y, kernel_width );
+
+    // apply horizontal smoothing to the vertically smoothed image to get a 2D
+    // box filter
+    box_average_horizontal( filter_x, filter_xy, kernel_width );
+
+    // Report the difference between the pixel value and all of the smoothed
+    // responses
+    vil_math_image_abs_difference( grey_img, filter_x, filter_x );
+    vil_math_image_abs_difference( grey_img, filter_y, filter_y );
+    vil_math_image_abs_difference( grey_img, filter_xy, filter_xy );
+
+    return output;
+  }
+
+  // Given an input grayscale image, and a smoothed version of this greyscale
+  // image, calculate the bidirectional filter response in the horizontal
+  // direction.
+  template < typename PixType >
+  void
+  horizontal_box_bidirectional_pass( const vil_image_view< PixType >& grey,
+                                     const vil_image_view< PixType >& smoothed,
+                                     vil_image_view< PixType >& output,
+                                     unsigned kernel_width )
+  {
+    const unsigned ni = grey.ni();
+    const unsigned nj = grey.nj();
+
+    output.fill( 0 );
+
+    unsigned offset = ( kernel_width / 2 ) + 1;
+
+    unsigned diff1 = 0, diff2 = 0;
+
+    {
```

Comment:
What is the purpose of this new scope here?  It shouldn't break anything but it looks extraneous.